File: src/types/dataset.ts
/**
 * TypeScript types for dataset API endpoints.
 */

/** Information about a single dataset within an HDF5 file. */
export interface HDF5DatasetInfo {
  path: string
  shape: number[]
  dtype: string
  is_4d: boolean
}

/** Response for single-datacube files (.dm4, .mrc). */
export interface SingleProbeResponse {
  type: 'single'
  shape: number[]
  dtype: string
}

/** Response for HDF5 files with a tree structure. */
export interface HDF5TreeProbeResponse {
  type: 'hdf5_tree'
  datasets: HDF5DatasetInfo[]
}

/** Union type for probe endpoint responses. */
export type ProbeResponse = SingleProbeResponse | HDF5TreeProbeResponse

/** Response from the dataset load endpoint. */
export interface DatasetLoadResponse {
  shape: number[]
  dtype: string
  dataset_path: string | null
}

/** Extended dataset info including file path and internal path. */
export interface DatasetInfo extends DatasetLoadResponse {
  filePath: string
}

/** Response from the mean diffraction endpoint. */
export interface MeanDiffractionResponse {
  image_base64: string
  width: number
  height: number
}

/** Response from the virtual image endpoint. */
export interface VirtualImageResponse {
  image_base64: string
  width: number
  height: number
}

/** Response from the diffraction pattern endpoint. */
export interface DiffractionPatternResponse {
  image_base64: string
  width: number
  height: number
  x: number
  y: number
}

/** Selection tool types for region selection. */
export type SelectionTool = 'rectangle' | 'ellipse' | 'lasso' | null

/** Region types for the API. */
export type RegionType = 'rectangle' | 'ellipse' | 'polygon'

/** Selection geometry stored in state. */
export interface SelectionGeometry {
  type: RegionType
  points: [number, number][]
}

/** Request for region-based diffraction pattern. */
export interface RegionDiffractionRequest {
  mode: 'mean' | 'max'
  region_type: RegionType
  points: [number, number][]
  log_scale?: boolean
  contrast_min?: number
  contrast_max?: number
}

/** Response from the region diffraction endpoint. */
export interface RegionDiffractionResponse {
  image_base64: string
  width: number
  height: number
  mode: string
  region_type: string
  pixels_in_region: number
}

File: backend/app/schemas/dataset.py
"""Dataset-related Pydantic schemas."""

from typing import Literal

from pydantic import BaseModel


class DatasetProbeRequest(BaseModel):
    """Request model for probing a dataset file."""

    path: str


class HDF5DatasetInfo(BaseModel):
    """Information about a single dataset within an HDF5 file."""

    path: str
    shape: list[int]
    dtype: str
    is_4d: bool


class HDF5TreeNode(BaseModel):
    """A node in the HDF5 file tree (group or dataset)."""

    name: str
    type: Literal["group", "dataset"]
    path: str
    children: list["HDF5TreeNode"] | None = None
    # Dataset-specific fields (only present when type == "dataset")
    shape: list[int] | None = None
    dtype: str | None = None
    is_4d: bool | None = None


# Enable self-referencing for recursive model
HDF5TreeNode.model_rebuild()


class SingleProbeResponse(BaseModel):
    """Response for single-datacube files (.dm4, .mrc)."""

    type: Literal["single"]
    shape: list[int]
    dtype: str


class HDF5TreeProbeResponse(BaseModel):
    """Response for HDF5 files with a hierarchical tree structure."""

    type: Literal["hdf5_tree"]
    filename: str
    root: HDF5TreeNode
    # Keep flat list for backwards compatibility
    datasets: list[HDF5DatasetInfo]


class DatasetLoadRequest(BaseModel):
    """Request model for loading a dataset."""

    path: str
    dataset_path: str | None = None


class DatasetLoadResponse(BaseModel):
    """Response model for loaded dataset metadata."""

    shape: list[int]
    dtype: str
    dataset_path: str | None = None
